Intra-node GPU messaging #4: Add example programs and CI file syntax highlighting

minitu opened the pull request:
Follow-up to #2903, last out of 4 PRs for intra-node GPU messaging.

- Added example programs that use the intra-node GPU messaging feature
- Added `nocopydevice` keyword to CI file syntax highlighting in Vim

PR sequence: #2896 -> #2901 -> #2903 -> #2904

nitbhat commented on examples/charm++/cuda/direct/jacobi2d/jacobi2d.C:
```diff
@@ -0,0 +1,409 @@
+    }
+  }
+
+  void receiveGhostsZC(int ref, int dir, int &w, double *&buf, CkDeviceBufferPost *devicePost) {
```
Since this will be an example, worth adding a comment here saying this is the "Post Entry method and the matching regular entry method is in the .ci file"

minitu replied:
👍

minitu commented on examples/charm++/cuda/direct/verify/verify.C:
```diff
@@ -0,0 +1,240 @@
+    hapiCheck(cudaStreamSynchronize(stream));
+
+    for (int i = 0; i < block_size; i++) {
+      if (h_remote_data[i] != val) {
```
👍

minitu commented on examples/charm++/cuda/direct/verify/verify.C:
```diff
@@ -0,0 +1,240 @@
+    nodegroup_proxy = CProxy_VerifyNodeGroup::ckNew();
+
+    // Begin testing
+    thisProxy.test();
```
👍

minitu replied:
Actually `test()` is a threaded entry method and needs to be invoked through the proxy.

minitu commented on examples/charm++/cuda/direct/pingpong/pingpong.C:
```diff
@@ -0,0 +1,260 @@
+          break;
+        case 'z':
+          use_zerocopy = true;
+          break;
```
👍

minitu commented on examples/charm++/cuda/direct/pingpong/pingpong.C:
```diff
@@ -0,0 +1,260 @@
+  void initDone() {
+    CkPrintf("Data initialized, starting test...\n");
+    cur_count = min_count;
+    thisProxy.testBegin(cur_count);
```
👍

minitu commented on examples/charm++/cuda/direct/pingpong/pingpong.C:
```diff
@@ -0,0 +1,260 @@
+        n_iters, use_zerocopy, validate);
+
+    // Create block group chare
+    block_proxy = CProxy_Block::ckNew(CkNumPes());
```
Oh I think this is an artifact of implementing support for chare arrays first. Will change this to a group

minitu commented on examples/charm++/cuda/direct/sdag/sdag.ci:
```diff
@@ -0,0 +1,43 @@
+
+        // Continue next iteration or return to Main for termination
+        if (iter++ < n_iters) {
+          thisProxy[thisIndex].iterate();
```
Calls to SDAG entry methods need to be done with the proxy, otherwise it will throw a runtime error.

minitu commented on examples/charm++/cuda/direct/sdag/sdag.C:
```diff
@@ -0,0 +1,153 @@
+            h_remote_data[i], (double)peer);
+        validated = false;
+      }
+      if (reg_remote_data[i] != peer) {
```
👍

minitu commented on examples/charm++/cuda/direct/sdag/sdag.C:
```diff
@@ -0,0 +1,153 @@
+    // Validate data
+    bool validated = true;
+    for (int i = 0; i < block_size; i++) {
+      if (h_remote_data[i] != (double)peer) {
```
👍

minitu commented on examples/charm++/cuda/direct/pingpong/pingpong.C:
```diff
@@ -0,0 +1,260 @@
+  void receive_zc(int count, double *data) {
+    // Wait for data transfer to complete
+    cudaStreamSynchronize(stream);
+
```
That's the difference from the host API, the regular entry method is invoked only after the data transfer is **initiated**.

minitu commented:
@nitbhat I did intend the lassen scripts to be included, so that I can use it later for benchmarking. I agree about leaving only `jacobi2d` and `sdag` in the examples folder. I'll let you know after I make the necessary changes.

minitu commented:
@nitbhat Addressed your comments. Moved `pingpong` to `benchmarks`, but leaving `verify`, `sdag`, and `jacobi2d` in examples for now as we currently aren't able to test CUDA programs with GitHub CI. Will move them to `tests` once we have the functionality.

nitbhat commented:
> All changes look good.
> 
> My only concern is the API semantics (esp since we're planning to merge the host and device API someday), I am wondering if different semantics will not allow that unification. We could change the API semantics then, but that should be the last resort as existing users will have to modify their already written code.
> 
> @minitu:
> You mentioned over slack that you can have `hapiAddCallback` in the RTS to ensure that the regular entry method will be invoked after the GPU data transfer is complete. But, I'm guessing the only reason you're not doing that is because you want to take advantage of offloading kernels that depend on the incoming data (and use the same stream?). Is that correct?

Just spoke to @minitu over slack and found out that ^ feature will be implemented in a separate PR and he's opened this issue (https://github.com/UIUC-PPL/charm/issues/2975) for it.

minitu commented:
@nitbhat @stwhite91 Just discarded changes to xi-scan.C which I think trickled over from the 3rd PR, Please approve again
